sox/transform.py

```diff
@@ -454,6 +461,102 @@ def build(self, input_filepath, output_filepath, extra_args=None,
             else:
                 return True
 
+    def build_array(self, input_array, sample_rate_in, sample_rate_out=None,
+                    channels_in=None, channels_out=None, bits_in=None, 
+                    bits_out=None, encoding_out=None, extra_args=None,
+                    return_output=True):
+        '''Builds an output numpy array by executing the current set of 
+        commands. Returns a tuple containing the status, the output array,
+        and the stderr: (status, output_array, stderr).
+
+        Parameters
+        ----------
+        input_array : np.ndarray
+            Input audio as a numpy array - (samples, channels).
+        sample_rate_in : int
+            Sample rate of audio. Since the data is given as a numpy 
+            array, we need this at build to set the input format correctly.
+        sample_rate_out : int, default=None
+            Expected sample rate of output audio. If None, this is set to
+            sample_rate_in.
+        channels_in : int, default=None
+            Number of channels in input. If None, defaults to number of channels
+            in input_array.
+        channels_out : int, default=None
+            Expected number of channels in output audio. If None, this is set 
+            to channels_in.
+        bits_in : int, default=None
+            Bitdepth of input audio. If None, inferred at runtime by sox. 
+        bits_out : int, default=None
+            Expected bitdepth of output audio. If None, inferred by sox.
+        encoding_out : str or None, default=None
+            str determining expected encoding of output audio. Must be one of
+            's8', 's16', 'f32', 'f64'.
+        extra_args : list or None, default=None
+            If a list is given, these additional arguments are passed to SoX
+            at the end of the list of effects.
+            Don't use this argument unless you know exactly what you're doing!
+        '''
+        output_filepath = '-'
+        input_filepath = '-'
+
+        if channels_in is None:
+            channels_in = (
+                input_array.shape[-1] if len(input_array.shape) > 1 else 1
+            )
+        if channels_out is None:
+            channels_out = channels_in
+        
+        if sample_rate_out is None:
+            sample_rate_out = sample_rate_in
+
+        encoding = input_array.dtype.type
+        if encoding_out is None:
+            encoding_out = encoding
+
+        self.set_input_format(
```

Comment:
I'm not sure we want to call these functions inside build_array - they change the internal state of the Transformer. The way it's written now, calling `build_array` will silently change the behavior of `build`.

Brainstorming how to get around this, we could:
1. write a function that builds the args list for input/output formats, which gets called by `set_input_format`/`set_output_format` and by this function. 
2. reduce the number of inputs to this function. as far as I can tell, the only needed input format argument is `sample_rate_in`. The rest can (should!?) be inferred from the input array. The output format support is already built into sox.build in several ways, either by calling `set_output_format` or by transformer commands like `rate` and `channels`. The only one I'd leave optionally specified here is `encoding_out`.

so, in short this function would become
```
build_array(self, input_array, sample_rate_in, encoding_out=None, 
            extra_args=None,  return_output=True)
```

Thoughts/objections?

Reply:
So the argument list for `build_array` was initially what you said, but had to be expanded quite a bit to get through all of the test cases. The ones I can get rid of are `channels_in` and `bits_in` as they can be inferred from the numpy array. But the output types can't be inferred. 

What if we had `set_output_format` and `set_input_format` return (optionally) the arg list rather than set it as a variable in the object? Then we could extend it dynamically in `build_array` instead without side effects.

Reply:
I modified `set_[input, output]_format` with a flag `return_only` where the arg list will be built without saving it to `self`. I had to keep some of the arguments to ` build_array`, but as you'll see further down, `build_array` is now merged with `build`.